cirq/circuits/circuit.py

```diff
@@ -1392,6 +1392,14 @@ def __pow__(self, exponent: int) -> 'Circuit':
 
     __hash__ = None  # type: ignore
 
+    def _with_measurement_key_mapping_(self, key_map: Dict[str, str]):
```

Comment:
it feels like this  and the FrozenCircuit version could be deduped into `AbstractCircuit` and use type(self) to instantiate?

Reply:
Done.

Reply:
Discovered why we can't dedupe these: `AbstractCircuit` has no `__init__` method, so mypy doesn't like it.

Reply:
OK. Alternative solutions:
1. We could create an @abstractmethod __init__ - not sure I like that
2. We could have a `with_operations` abstract method that would be implemented in the subclasses as `Circuit(ops, device=self.device`.
WDYT?

Reply:
Option (2) seems right. I'm also realizing that `Circuit(circuit.ops)` might not preserve the moment structure of the original circuit, so we may want to reuse the existing `_with_sliced_moments` method for this.

Reply:
That is a great point - this whole PR is missing remapping a Moment!!! That's the definitely the level where we should operate.

Reply:
Protocol method is now provided for moments, and tests are added to ensure that Circuits correctly preserve the moment structure during remapping.